### ISSUE TITLE
Migration fails due to theme

### ISSUE THREAD

**Issue by revant** [frappe/frappe_docker#695]
```
Traceback (most recent call last):
  File "/usr/local/bin/patched_bench_helper.py", line 46, in <module>
    raise SystemExit(main())
  File "/usr/local/bin/patched_bench_helper.py", line 41, in main
    frappe.utils.bench_helper.main()
  File "/home/frappe/frappe-bench/apps/frappe/frappe/utils/bench_helper.py", line 19, in main
    click.Group(commands=commands)(prog_name='bench')
  File "/home/frappe/frappe-bench/env/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/frappe/frappe-bench/env/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/frappe/frappe-bench/env/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/frappe/frappe-bench/env/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/frappe/frappe-bench/env/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/frappe/frappe-bench/env/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/frappe/frappe-bench/env/lib/python3.9/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/commands/__init__.py", line 27, in _func
    ret = f(frappe._dict(ctx.obj), *args, **kwargs)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/commands/site.py", line 309, in migrate
    migrate(
  File "/home/frappe/frappe-bench/apps/frappe/frappe/migrate.py", line 91, in migrate
    frappe.get_attr(fn)()
  File "/home/frappe/frappe-bench/apps/frappe/frappe/website/doctype/website_theme/website_theme.py", line 187, in after_migrate
    doc.generate_bootstrap_theme()
  File "/home/frappe/frappe-bench/apps/frappe/frappe/website/doctype/website_theme/website_theme.py", line 80, in generate_bootstrap_theme
    frappe.throw('<div style="font-family: monospace;">{stderr}</div>'.format(stderr=stderr))
  File "/home/frappe/frappe-bench/apps/frappe/frappe/__init__.py", line 444, in throw
    msgprint(msg, raise_exception=exc, title=title, indicator='red', is_minimizable=is_minimizable, wide=wide, as_list=as_list)
  File "/home/frappe/frappe-bench/apps/frappe/frappe/__init__.py", line 423, in msgprint
    _raise_exception()
  File "/home/frappe/frappe-bench/apps/frappe/frappe/__init__.py", line 378, in _raise_exception
    raise raise_exception(msg)
frappe.exceptions.ValidationError: <div style="font-family: monospace;">internal/modules/cjs/loader.js:905<br>  throw err;<br>  ^<br><br>Error: Cannot find module 'node-sass'<br>Require stack:<br>- /home/frappe/frappe-bench/apps/frappe/generate_bootstrap_theme.js<br>    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:902:15)<br>    at Function.Module._load (internal/modules/cjs/loader.js:746:27)<br>    at Module.require (internal/modules/cjs/loader.js:974:19)<br>    at require (internal/modules/cjs/helpers.js:101:18)<br>    at Object.<anonymous> (/home/frappe/frappe-bench/apps/frappe/generate_bootstrap_theme.js:1:14)<br>    at Module._compile (internal/modules/cjs/loader.js:1085:14)<br>    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)<br>    at Module.load (internal/modules/cjs/loader.js:950:32)<br>    at Function.Module._load (internal/modules/cjs/loader.js:790:12)<br>    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:75:12) {<br>  code: 'MODULE_NOT_FOUND',<br>  requireStack: [<br>    '/home/frappe/frappe-bench/apps/frappe/generate_bootstrap_theme.js'<br>  ]<br>}<br></div>

```

This only happens for site which has `Website Theme` applied.

@vrslev I think we need to make node-sass available in erpnext-python somehow.

**Comment by vrslev** [comment]
This is relevant only to _ERPNext_ images?

**Comment by revant** [comment]
It exists in frappe image.

1. Create a `Website Theme`
2. Set the created theme in `Website Settings`
3. migrate the site.

**Comment by vrslev** [comment]
Did it work before refactor? 🤔

**Comment by revant** [comment]
yes. there was some node_module gymnastics.

If we place node_modules from the nginx image in /home/frappe/frappe-bench/apps/frappe in worker image then It will work.

In helm chart I'm planning to mount or copy them some how.

**Comment by vrslev** [comment]
> If we place node_modules from the nginx image in /home/frappe/frappe-bench/apps/frappe in worker image then It will work.

Should we and how can we do that?

**Comment by revant** [comment]
How to populate node_modules? 

should the be part of python image?

or should we populate like this?

What happens to custom apps with node_modules? App makers decision?

**Comment by vrslev** [comment]
I don't know. Generating Website Theme in web with calling node from subprocess is such bad design 😅

**Comment by vrslev** [comment]
It definitely shouldn't be a part of backend image. I wonder how it worked before

**Comment by revant** [comment]
> I wonder how it worked before

I think previous image had entrypoint that set the symlink to node_modules from assets volume

node-sass in production node_modules is also not done in other projects.

**Comment by vrslev** [comment]
> node-sass in production node_modules is also not done in other projects. 

Don't understand what you're trying to say here

**Comment by revant** [comment]
> I don't know. Generating Website Theme in web with calling node from subprocess is such bad design sweat_smile

we can just make an issue in frappe/frappe for discussion. Generally if it work

> Don't understand what you're trying to say here

Bad design patterns is a different can of worms. Let it be closed. 

> We already expose assets to backend image. Maybe we can symlink them in new setup, too?

symlink needs to happen during runtime. It was in entrypoint.

https://github.com/frappe/frappe_docker/blob/95aeb32e2d1d3d5754bf4f97d8af650d462488d1/build/frappe-worker/docker-entrypoint.sh#L71-L72

**Comment by vrslev** [comment]
> Bad design patterns is a different can of worms. Let it be closed.

True.

We can add entrypoint.sh for backend image with following contents:
```bash
#!/bin/bash

set -e

# Link frappe node_modules for Website Theme
if test -d /home/frappe/frappe-bench/sites/assets/frappe/node_modules && test -d /home/frappe/frappe-bench/apps/frappe/node_modules; then
  ln -sfn /home/frappe/frappe-bench/sites/assets/frappe/node_modules /home/frappe/frappe-bench/apps/frappe/node_modules
fi

exec "$@"
```
and use it as entrypoint.
What do you think?

**Comment by revant** [comment]
Can it be like this pseudo code:

```
if /home/frappe/frappe-bench/assets/{app_name}/node_modules exists:
   ln -sfn /home/frappe/frappe-bench/assets/{app_name}/node_modules /home/frappe/frappe-bench/apps/{app_name}/node_modules
```

**Comment by vrslev** [comment]
I think it is too much overhead 🤔

**Comment by vrslev** [comment]
In theory, sure. If we assume every directory in /home/frappe/frappe-bench/assets is an app

**Comment by revant** [comment]
then we let the app maker override the entrypoint if their app has node_modules with "bad design"?

**Comment by revant** [comment]
> and use it as entrypoint. What do you think?

This is fine

**Comment by vrslev** [comment]
> then we let the app maker override the entrypoint if their app has node_modules with "bad design"?

Maybe, do nothing with custom apps until somebody leaves an issue? I'm sure the thing we are discussing here is not that common.

> This is fine

Can you add an PR?

**Comment by revant** [comment]
> Maybe, do nothing with custom apps until somebody leaves an issue?

override entrypoint.sh, just create your own script and copy it over the one in official image! Solved for custom apps with bad design. Why use bad design pattern in custom app where you have choice not to? If it is still done then overriding the script is also easy thing to do.

> Can you add an PR?

Sure. I'll just do things mentioned in https://github.com/frappe/frappe_docker/issues/695#issuecomment-1067030962, test and send PR.

**Comment by vrslev** [comment]
Resolved in #702
